ui-modules/blueprint-composer/app/components/util/model/entity.model.js

```diff
@@ -685,7 +685,17 @@ function addConfigKeyDefinition(param, overwrite, skipUpdatesDuringBatch) {
         let key = (param || {}).name;
         if (!key) throw new Error("'name' field must be included when adding parameter; was", param);
 
-        allConfig[key] = Object.assign(allConfig[key] || {}, param, overwrite ? null : allConfig[key]);
+        let paramMapped = Object.assign({}, param);
```

Comment:
I feel confident the below would be a much cleaner version of this function with the same behaviour. I'd really suggest something close for maintenance purposes if it feels accurate logic-wise.

```javascript
function addConfigKeyDefinition(param, overwrite, skipUpdatesDuringBatch) {
    const allConfig = this.miscDataOrDefault('configMap', {});
    
    if (typeof param !== 'undefined') {
        
        const parameterItem = (typeof param !== 'string')
            ? { ...param } // shallow copy as object
            : {
                "name": param,
                "label": param,
                "description": "",
                "priority": 1,
                "pinned": true,
                "type": "java.lang.String",
                "constraints": [],
            };

        const key = get(parameterItem, 'name'); // import { get } from 'lodash' at top, defaults to undef.

        // this will also throw if original `param` is neither object nor string due to previous statement
        if (!key) throw new Error("'name' field must be included when adding parameter; was", param);

        if (typeof parameterItem.default !== 'undefined') {
            /* Annoyingly, in parameters, we call the default value "default",
             * but in config, we call them "defaultValue";
             * when params are merged to config we need to rename
             */
            parameterItem.defaultValue = parameterItem.default;
            delete parameterItem.default;
        }

        Object.assign(allConfig[key], parameterItem);
    }

    if (!skipUpdatesDuringBatch) {
        this.miscData.set('config', Object.values(allConfig));
    }

    this.touch();
    return this;
}
```

Reply:
slightly clearer and more efficient apart from (1) you've lost the `overwrite` behaviour and (2) bringing in `get` obscures behaviour for those not familiar with `_.get` -- but the benefit is so slight it isn't worth the effort of testing it

Reply:
Yeah fair, just though I'd share. For the record, `overwrite` was removed because it seemed to contribute no different logic path to the end result due to the syntax I went with that allowed a simpler extension.

Reply:
if `overwrite` was false or set false because `param` was supplied as a string, then the difference is that the code should do `Object.assign(parameterItem, allConfig[key])` prior to the `Object.assign(allConfig[key], parameterItem)` -- the difference being if both set a value for the same key, which one should be kept.  With this change any key set in the param object will always overwrite such a key in config.